trie-db/src/lookup.rs

```diff
@@ -14,66 +14,533 @@
+			.flatten()
+		{
+			hash
+		} else {
+			let hash_and_value = self.look_up_with_cache_internal(
+				nibble_key,
+				full_key,
+				cache,
+				|value, _, full_key, _, _, recorder| match value {
+					ValueOwned::Inline(value, hash) => {
+						if let Some(recorder) = recorder.as_mut() {
+							recorder.record(TrieAccess::Value {
+								hash,
+								value: value.as_ref().into(),
+								full_key,
+							});
+						}
+
+						Ok((hash, Some(value.clone())))
+					},
+					ValueOwned::Node(hash) => {
+						if let Some(recoder) = recorder.as_mut() {
+							recoder.record(TrieAccess::Hash { full_key });
+						}
+
+						Ok((hash, None))
+					},
+				},
+			)?;
+
+			match &hash_and_value {
+				Some((hash, Some(value))) =>
+					cache.cache_value_for_key(full_key, (value.clone(), *hash).into()),
+				Some((hash, None)) => cache.cache_value_for_key(full_key, (*hash).into()),
+				None => cache.cache_value_for_key(full_key, CachedValue::NonExisting),
+			}
+
+			hash_and_value.map(|v| v.0)
+		};
+
+		Ok(res)
+	}
+
 	/// Look up the given key. If the value is found, it will be passed to the given
 	/// function to decode or copy.
-	pub fn look_up(mut self, key: NibbleSlice) -> Result<Option<Q::Item>, TrieHash<L>, CError<L>> {
-		let mut partial = key;
+	///
+	/// It uses the given cache to speed-up lookups.
+	fn look_up_with_cache(
+		mut self,
+		full_key: &[u8],
+		nibble_key: NibbleSlice,
+		cache: &mut dyn crate::TrieCache<L::Codec>,
+	) -> Result<Option<Q::Item>, TrieHash<L>, CError<L>> {
+		let trie_nodes_recorded =
+			self.recorder.as_ref().map(|r| r.trie_nodes_recorded_for_key(full_key));
+
+		let (value_cache_allowed, value_recording_required) = match trie_nodes_recorded {
+			// If we already have the trie nodes recorded up to the value, we are allowed
+			// to use the value cache.
+			Some(RecordedForKey::Value) | None => (true, false),
+			// If we only have recorded the hash, we are allowed to use the value cache, but
+			// we may need to have the value recorded.
+			Some(RecordedForKey::Hash) => (true, true),
+			// As we don't allow the value cache, the second value can be actually anything.
+			Some(RecordedForKey::None) => (false, true),
+		};
+
+		let res = match value_cache_allowed.then(|| cache.lookup_value_for_key(full_key)).flatten()
+		{
+			Some(CachedValue::NonExisting) => None,
+			Some(CachedValue::ExistingHash(hash)) => {
+				let data = Self::load_owned_value(
+					// If we only have the hash cached, this can only be a value node.
+					// For inline nodes we cache them directly as `CachedValue::Existing`.
+					ValueOwned::Node(*hash),
+					nibble_key.original_data_as_prefix(),
+					full_key,
+					cache,
+					self.db,
+					&mut self.recorder,
+				)?;
+
+				cache.cache_value_for_key(full_key, data.clone().into());
+
+				Some(data.0)
+			},
+			Some(CachedValue::Existing { data, hash, .. }) =>
+				if let Some(data) = data.upgrade() {
+					if value_recording_required {
```

Comment:
I checked this a bit more, it is indeed a bit of work, but makes things a lot more straight forward.
Having a single non optional trait can make things way more systematic.

So sorry to bring the idea again, but I think it could make things simpler.

- trie node cache and recording: here any call to db.get should be always inside `get_or_insert_node`.
Having caching or recording active and inactive is simply a choice of the trait implementation used when building triedb or triedbmut. eg https://github.com/cheme/trie/blob/dc36257fd7dd043d39e1de92584b8f193875c7f0/trie-db/src/lib.rs#L497 the recording and cache works together ok.
So caching and recording is only done when hitting this `db.get`. Exception made for `cache_node` in triedbmut that can still use `get_or_insert_node` but without hitting the db.

- value cache and recording: can be only call from top level function like 'lookup' and 'lookup_hash'.
It would be better if `lookup_value_for_key` and `cache_value_for_key` are only called from top level lookup function. This may not work well with the WeakBytes reference system (except if we change api to allow return &Byte from cache). At least when caching missing value, the caching could be done at the top level in front of the `lookup` call.

Having a single trait for it that is not optional also allows removing all redundancy from lookup.rs.
When checking this I also realize that value could use their own primitives, also doing so means that both OwnedNode and NodeOwned need to be cache (I did some test in https://github.com/cheme/trie/tree/bkchr-funny-branch).

Things such as checking in `lookup_value` if we did already access the value can be just part of the implementation as in https://github.com/cheme/trie/blob/dc36257fd7dd043d39e1de92584b8f193875c7f0/trie-db/src/lib.rs#L463

Reply:
I try to just put Option<&mut dyn Recorder> and Option<&mut dyn TrieCache> under a shared struct to allow same thing:
- consistently use get_or_insert_node when accessing any trie_node
- have interaction between recorder and cache done behind this struct implementation only.
Basically same attempt as previously but not using trait to stay closer to this branch code.

I quickly realize, that having two different owned node implementation (triedbmut one and the new node one) was an issue.

So I tried merging both in https://github.com/paritytech/trie/compare/bkchr-funny-branch...cheme:bkchr-funny-branch3?expand=1.
I am not exactly happy with the result (even though some redundant code got remove): there is some unreachable code due to triedbmut storage being unreachable from triedb.
Clearly the issue is this triedbmut node store that only helps avoiding using a bit of unsafe code. Still I don't think removing it and use unsafe code instead is worth it.